### ISSUE TITLE
Homebrew CMake build doesn't include documentation

### ISSUE THREAD

**Issue by thomcc** [fish-shell/fish-shell#4809]
Versions and such.

- `fish --version`: `fish, version 2.7.1-971-g0f6c763d.`
- `uname -a`: `Darwin void 17.3.0 Darwin Kernel Version 17.3.0: Thu Nov  9 18:09:22 PST 2017; root:xnu-4570.31.3~1/RELEASE_X86_64 x86_64` (macOS 10.13.2).
- `echo $TERM`: `xterm-256color` (but I'm using iTerm2 Build 3.1.5)

Also, not requested but could be relevant:

- `brew --version`: `Homebrew 1.5.10-10-g144356e`
- `cmake --version`: `cmake version 3.9.0`

I just installed fish HEAD via homebrew. It doesn't have documentation, `ls /usr/local/Cellar/fish/HEAD-0f6c763/share/doc/fish/` shows just `CHANGELOG.md`. We don't seem to have the manpages in `/usr/local/Cellar/fish/HEAD-0f6c763/share/fish/man/man1` either (in fact, `/usr/local/Cellar/fish/HEAD-0f6c763/share/fish/man` does not exist), so `__fish_print_commands` doesn't work, and so help falls through [here](https://github.com/fish-shell/fish-shell/blob/master/share/functions/help.fish#L98), meaning it only ever opens the documentation's index (and so is pretty annoying to use, since it doesn't even really give me outdated information for a command).

Undoing this change in homebrew: https://github.com/Homebrew/homebrew-core/commit/a3623730b0fca7294c4d285dcfe8ed64e1261ba3 fixes basically all of this (after removing the broken version, and reinstalling), but my suspicion is that this is an issue with our CMake build rather than the formula, so I filed here instead of there. I could completely be wrong about this though!

I haven't noticed anything else broken, but I happened to notice help being busted pretty early, and so I dug into it (initially I thought it was an issue with the URL trampoline code I wrote for `help` recently, but thankfully it's not my fault that `help` is broken :p). Regardless, my point is it could easily have failed to install things other than the docs/manpages as well, I just wouldn't have noticed.

**Comment by faho** [comment]
I have also hit this issue on Linux.

I think the problem is that the documentations aren't installed correctly, because we missed a `${CMAKE_CURRENT_BINARY_DIR}`.

I tried fixing this in e800b18e, and I accidentally pushed it (that's what happens when you add experimental commits to your master branch).

So this might actually be fixed.

**Comment by zx8** [comment]
@faho Unfortunately not fixed with Homebrew…

```sh
$ brew reinstall fish --HEAD

[...]

Last 15 lines from /Users/zx8/Library/Logs/Homebrew/fish/02.make:
[100%] Linking CXX executable fish
/usr/local/Cellar/cmake/3.10.2/bin/cmake -E cmake_link_script CMakeFiles/fish.dir/link.txt --verbose=1
/usr/local/Homebrew/Library/Homebrew/shims/super/clang++  -DNDEBUG -Wl,-search_paths_first -Wl,-headerpad_max_install_names  CMakeFiles/fish.dir/src/fish.cpp.o  -o fish libfishlib.a libtinyexpr.a /usr/lib/libcurses.dylib /usr/local/lib/libpcre2-32.dylib
[100%] Built target fish
/usr/local/Cellar/cmake/3.10.2/bin/cmake -E cmake_progress_start /tmp/fish-20180313-68243-1x1cnwg/CMakeFiles 0
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f CMakeFiles/Makefile2 preinstall
make[1]: Nothing to be done for `preinstall'.
Install the project...
/usr/local/Cellar/cmake/3.10.2/bin/cmake -P cmake_install.cmake
-- Install configuration: "Release"
CMake Error at cmake_install.cmake:178 (file):
  file INSTALL cannot find "/tmp/fish-20180313-68243-1x1cnwg/user_doc/html".
```

**Comment by zanchey** [comment]
e800b18 broke the build for all PPAs and OBS platforms. I reverted it as b5aea3fd.

**Comment by zanchey** [comment]
This is a problem for in-tree builds only; I'll take a look tomorrow but it's important to keep the matrix of in-tree/out-of-tree / with-Doxygen/tarball-without-Doxygen working.

**Comment by faho** [comment]
>e800b18 broke the build for all PPAs and OBS platforms

Terribly sorry about that! That wasn't supposed to be pushed!

>This is a problem for in-tree builds only

No - this happens for my out-of-tree builds as well. Or am I missing a step?

I'm doing

```shell
        cmake -G Ninja "$srcdir/$_gitname" \
              -DCMAKE_INSTALL_PREFIX=/usr \
              -DCMAKE_INSTALL_DATADIR=share \
              -DCMAKE_INSTALL_DOCDIR=share/doc/fish \
              -DCMAKE_INSTALL_SYSCONFDIR=/etc
        ninja
        ninja doc
        DESTDIR="$pkgdir" ninja install
```

and I get the docs in a "user_doc/html" directory in the build directory, but they aren't installed into "$pkgdir".

**Comment by zanchey** [comment]
I just merged a whole bunch of changes that I think finally bring the CMake build for in-tree, out-of-tree, tarball in-tree and tarball out-of-tree builds into alignment with the existing build system. I get documentation in my Homebrew installation now - is it working for others?

**Comment by zx8** [comment]
@zanchey Seems good now, apart from:

```
$ help string
# Opens browser @ file:///usr/local/Cellar/fish/HEAD-c0535b4/share/doc/fish/commands.html#string
```
![image](https://user-images.githubusercontent.com/8544900/37867389-7cd9c4aa-2f8f-11e8-894d-c0a8497d25b4.png)

**Comment by zanchey** [comment]
@zx8, that got fixed in 62870abe342ab7e38dbfe76e6b7f0569d371af29.
